opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -87,23 +87,50 @@ message InstrumentationLibraryMetrics {
 //                       +---------------------------+
 //
 //-----------------------------------------------------------------------
-// DataPoint is a value of specific type corresponding to a given moment in
-// time. Each DataPoint is timestamped.
-//
-// DataPoint is strongly typed: each DataPoint type has a specific Protobuf message
-// depending on the value type of the metric and thus there are currently 4 DataPoint
-// messages, which correspond to the types of metric values.
 message Metric {
   // metric_descriptor describes the Metric.
   MetricDescriptor metric_descriptor = 1;
 
-  // Data is a list of one or more DataPoints for a single metric. Only one of the
-  // following fields is used for the data, depending on the type of the metric defined
-  // by MetricDescriptor.type field.
-  repeated Int64DataPoint int64_data_points = 2;
-  repeated DoubleDataPoint double_data_points = 3;
-  repeated HistogramDataPoint histogram_data_points = 4;
-  repeated SummaryDataPoint summary_data_points = 5;
+  repeated Data data = 2;
+
+  message Data {
+    // The set of labels that uniquely identify this timeseries.
+    //
+    // These labels take precedence over any labels with the same key defined
+    // in the metric descriptor.
+    repeated opentelemetry.proto.common.v1.StringKeyValue labels = 1;
+
+    // start_time_unix_nano is the beginning of a time interval over which
+    // this data point was recorded. It MUST be represented as the UNIX Epoch
+    // time in nanoseconds (nanoseconds since 00:00:00 UTC on 1 January 1970).
+    // If zero, it is treated as unspecified and may be set by the receiver.
+    //
+    // If this data point is a part of a timeseries for an INSTANTANEOUS
+    // metric, this value is ignored.
+    //
+    // If this data point is a part of a timeseries for an DELTA, or
+    // CUMULATIVE metric, this value represents the instant before the
+    // measuring time interval began (i.e. for time t to be in the measuring
+    // interval than start_time_unix_nano < t <= time_unix_nano).
+    //
+    // This time MUST be reset so as to represent a new measurement lifetime
+    // if this data point is part of a timeseries for a CUMULATIVE metric and
+    // the metric value is reset to zero.
+    fixed64 start_time_unix_nano = 2;
+
+    // time_unix_nano is the time when this data point was recorded. It MUST
+    // be represented as the UNIX Epoch time in nanoseconds (nanoseconds since
+    // 00:00:00 UTC on 1 January 1970). If zero, it is treated as unspecified
+    // and may be set by the receiver.
+    fixed64 time_unix_nano = 3;
+
+    // Value is the measured value.
+    oneof Value {
```

Comment:
`oneof` has a high cost, unfortunately. Either the recently-released (new) Go protobuf library can help or improve on this, otherwise I suggest we just remove it and document that only one of the three types should be present, defined accordingly in the metric descriptor.

Reply:
I agree and we did especially change it, so look also to the history of PRs for more discussion.